compiler/rustc_data_structures/src/lib.rs

```diff
@@ -131,6 +136,24 @@ impl<F: Fn()> Drop for OnDrop<F> {
     }
 }
 
+struct IterFromGenerator<G>(G);
+
+impl<G: Generator + Unpin> Iterator for IterFromGenerator<G> {
+    type Item = G::Yield;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match Pin::new(&mut self.0).resume(()) {
+            GeneratorState::Yielded(n) => Some(n),
+            GeneratorState::Complete(_) => None,
+        }
+    }
+}
+
+/// An adapter for turning a generator closure into an iterator, similar to `iter::from_fn`.
+pub fn iter_from_generator<G: Generator + Unpin>(generator: G) -> impl Iterator<Item = G::Yield> {
```

Comment:
I almost feel this should be in `std`.

Reply:
I'll make separate PR to add this to libcore, not sure whether it will be merged and how much time it may take.
I don't know why this functionality wasn't added yet, maybe there are reasons - the body of work produced by async-wg is vast, but I didn't read any of it.